src/window.c

```diff
@@ -412,6 +426,95 @@ void draw_border(node_t *n, bool focused_node, bool focused_monitor)
 	}
 }
 
+void window_rounded_border(node_t *n)
+{
+	xcb_window_t win = n->id;
+	unsigned int radius = n->client->drawn_border_radius;
+
+	if (n->client->sets_own_shape) return;
+
+	// get geometry
+	xcb_get_geometry_reply_t *geo = xcb_get_geometry_reply(dpy, xcb_get_geometry(dpy, win), NULL);
+	if (geo == NULL) return;
+
+	uint16_t x  = geo->x;
+	uint16_t y  = geo->y;
+	uint16_t w  = geo->width;
+	uint16_t h  = geo->height;
+	uint16_t bw = geo->border_width;
+	uint16_t ow  = w+2*bw;
+	uint16_t oh  = h+2*bw;
+
+	free(geo);
+
+	xcb_pixmap_t bpid = xcb_generate_id(dpy);
+	xcb_pixmap_t cpid = xcb_generate_id(dpy);
+
+	xcb_create_pixmap(dpy, 1, bpid, win, ow, oh);
+	xcb_create_pixmap(dpy, 1, cpid, win, w, h);
+
+	xcb_gcontext_t black = xcb_generate_id(dpy);
+	xcb_gcontext_t white = xcb_generate_id(dpy);
+
+	xcb_create_gc(dpy, black, bpid, XCB_GC_FOREGROUND, (uint32_t[]){0, 0});
+	xcb_create_gc(dpy, white, bpid, XCB_GC_FOREGROUND, (uint32_t[]){1, 0});
+
+	int32_t rad, dia;
+	rad = radius;
+
+	rad += bw; dia = rad*2-1;
+
+	xcb_arc_t barcs[] = {
+		{ -1,     -1,     dia, dia, 0, 360 << 6 },
+		{ -1,     oh-dia, dia, dia, 0, 360 << 6 },
+		{ ow-dia, -1,     dia, dia, 0, 360 << 6 },
+		{ ow-dia, oh-dia, dia, dia, 0, 360 << 6 },
+	};
+	xcb_rectangle_t brects[] = {
+		{ rad, 0, ow-dia, oh },
+		{ 0, rad, ow, oh-dia },
+	};
+
+	rad -= bw; dia = rad*2-1;
+
+	xcb_arc_t carcs[] = {
+		{ -1,    -1,    dia, dia, 0, 360 << 6 },
+		{ -1,    h-dia, dia, dia, 0, 360 << 6 },
+		{ w-dia, -1,    dia, dia, 0, 360 << 6 },
+		{ w-dia, h-dia, dia, dia, 0, 360 << 6 },
+	};
+	xcb_rectangle_t crects[] = {
+		{ rad, 0, w-dia, h },
+		{ 0, rad, w, h-dia },
+	};
+
+	xcb_rectangle_t bounding = {0, 0, w+2*bw, h+2*bw};
+	xcb_poly_fill_rectangle(dpy, bpid, black, 1, &bounding);
+	xcb_poly_fill_rectangle(dpy, bpid, white, 2, brects);
+	xcb_poly_fill_arc(dpy, bpid, white, 4, barcs);
+
+	xcb_rectangle_t clipping = {0, 0, w, h};
+	xcb_poly_fill_rectangle(dpy, cpid, black, 1, &clipping);
+	xcb_poly_fill_rectangle(dpy, cpid, white, 2, crects);
+	xcb_poly_fill_arc(dpy, cpid, white, 4, carcs);
+
+	xcb_shape_mask(dpy, XCB_SHAPE_SO_SET, XCB_SHAPE_SK_BOUNDING,  win, -bw, -bw, bpid);
+	xcb_shape_mask(dpy, XCB_SHAPE_SO_SET, XCB_SHAPE_SK_CLIP, win, 0, 0, cpid);
```

Comment:
Doing this will crop the window contents: I'm therefore not in favour of modifying the clip region.

Reply:
But there's no other way
The corners will have to be clipped for rounded corners.

I'm not sure what you have in mind as an alternative..

Reply:
If the clip region must be specified, it should just be the default rectangle (`{0, 0, width, height}`).

Reply:
that would either result in this:
![](https://i.imgur.com/9bZzakq.png)
or this:
![](https://i.imgur.com/NdKdQEr.png)

And rounded corners for people not using borders (e.g.: me) would be impossible

Reply:
I'm currently using this patch, and I'm one of those people without borders in the config (```bspc config border_width 0```).

For it to work nicely I configured my urxvt with ```URxvt.internalBorder: 20```, and this is the result:

![screenshot-2018-10-25t13 38 55-0300](https://user-images.githubusercontent.com/6091633/47516383-bb40a680-d85b-11e8-9fb2-16744838b504.png)

As you can see, it's working nicely... but I don't have an equivalent option for other programs e.g. Qutebrowser:

![screenshot-2018-10-25t13 46 47-0300](https://user-images.githubusercontent.com/6091633/47516669-6f423180-d85c-11e8-962e-d041f96a47de.png)


As you can see, the rounded corner is messing up with my scroll status bar, and in a few websites it may also mess up with the content. Yeah, I'm abusing this feature in favor of ＡＥＳＴＨＥＴＩＣＳ, but that's the whole idea, isn't it? 

So maybe both use cases may happen to other users, and maybe the "internal corner radius" should be configured separately, so it's possible to have an "internalBorder" equivalent for other programs in similar setups with ```border_width > 0```, while setups without borders (like my terminal + ```border_width  = 0```) will still be possible too.

Reply:
I was gonna suggest that yeah, (means more work for me though lol)
If we have inner radius and outer radius options separately, we also need an `auto` option for those who want to the have the inner radius nicely calculated for them like it is now (`inner_rad = outer_rad - border_width`).

So my suggestion is (and I'm trying to integrate the `border_radius` -> `corner_radius` rename here):
- `corner_radius`: outer corner radius in pixels. Default=0
- `inner_corner_radius`: inner corner radius in pixels OR `auto`. Default=auto

(I'm not too sure how I feel about the naming here...)

Reply:
> that would either result in this:
Having an inner radius of 0 would, of course, mean that the outer radius should be less or equal to the border width. And since most users set a small border width it would imply having a small corner width.